kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -313,52 +324,101 @@ private long commitWithRetry(
                 dataActions);
       }
 
-      int numTries = 0;
-      while (numTries <= maxRetries) { // leq because the first is a try, not a retry
-        logger.info("Committing transaction as version = {}.", commitAsVersion);
+      int attempt = 1;
+      boolean seenRetryableNonConflictException = false;
+      while (true) {
+        // This loop exits upon either (a) commit success (return statement) or (b) commit failure.
+        logger.info(
+            "Attempting to commit transaction at table version {}. Attempt {}/{}",
+            commitAsVersion,
+            attempt,
+            getMaxCommitAttempts());
         try {
           transactionMetrics.commitAttemptsCounter.increment();
           return doCommit(
               engine, commitAsVersion, attemptCommitInfo, dataActions, transactionMetrics);
-        } catch (FileAlreadyExistsException fnfe) {
-          logger.info(
-              "Concurrent write detected when committing as version = {}.", commitAsVersion);
-          if (numTries < maxRetries) {
-            // only try and resolve conflicts if we're going to retry
+        } catch (CommitFailedException cfe) {
+          if (!cfe.isRetryable()) {
+            // Case 1: Non-retryable exception. We must throw this. We don't expect connectors to
+            //         be able to recover from this.
+            throw new RuntimeException(
+                String.format(
+                    "Commit attempt %d for version %d failed with a non-retryable exception.",
+                    attempt, commitAsVersion),
+                cfe);
+          } else if (attempt >= getMaxCommitAttempts()) {
+            // Case 2: Despite the error being retryable, we have exhausted the maximum number of
+            //         retries. We must throw here, too.
+            throw new MaxCommitRetryLimitReachedException(commitAsVersion, maxRetries, cfe);
+          } else if (!cfe.isConflict()) {
+            // Case 3: No conflict => No conflict resolution needed. Just retry with same version.
+            logger.warn(
+                "Commit attempt {} for table version {} failed with a retryable exception and "
+                    + "without conflict. Skipping conflict resolution and trying again. "
+                    + "Exception: {}",
+                attempt,
+                commitAsVersion,
+                cfe);
+            seenRetryableNonConflictException = true;
+          } else if (seenRetryableNonConflictException) {
```

Comment:
moving the two conflict cases to a helper method might also make the flow easier to understand here.